pennylane/tape/functions/unified.py

```diff
@@ -0,0 +1,211 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Wrappers for common functions that manipulate or create
+NumPy, TensorFlow, and Torch data structures.
+"""
+import abc
+
+
+class UnifiedTensor(abc.ABC):
+    """Creates a unified tensor wrapper, that represents interface array/tensor datatypes
+    using a common API.
+
+    While this is an abstract base class, this class may be 'instantiated' directly;
+    by overloading ``__new__``, the tensor argument is inspected, and the correct subclass
+    is returned:
+
+    >>> x = tf.Variable([0.4, 0.1, 0.5])
+    >>> y = UnifiedTensor(x)
+    >>> type(y)
+    <class 'pennylane.tape.functions.tf.TensorFlowTensor'>
+
+    The original tensor is available via the ``data`` attribute.
+
+    In addition, this class defines various abstract methods that all subclasses
+    must define. These methods allow for common manipulations and
+    linear algebra transformations without the need for importing.
+
+    >>> y.ones_like()
+    tf.Tensor([1. 1. 1.], shape=(3,), dtype=float32)
+
+    Unless specified, the returned tensots are also unified tensors, allowing
+    for method chaining:
+
+    >>> y.ones_like().expand_dims(0)
+    tf.Tensor([[1. 1. 1.]], shape=(1, 3), dtype=float32)
+    """
+
+    def __new__(cls, tensor):
+        if isinstance(tensor, UnifiedTensor):
+            return tensor
+
+        if cls is not UnifiedTensor:
+            return super(UnifiedTensor, cls).__new__(cls)
+
+        if isinstance(tensor, (list, tuple)):
+            from .autograd import AutogradTensor, np
+
+            return AutogradTensor.__new__(AutogradTensor, np.array(tensor))
+
+        namespace = tensor.__class__.__module__.split(".")[0]
+
+        if namespace in ("pennylane", "autograd", "numpy"):
+            from .autograd import AutogradTensor
+
+            return AutogradTensor.__new__(AutogradTensor, tensor)
+
+        if namespace == "tensorflow":
+            from .tf import TensorFlowTensor
+
+            return TensorFlowTensor.__new__(TensorFlowTensor, tensor)
+
+        if namespace == "torch":
+            from .torch import TorchTensor
+
+            return TorchTensor.__new__(TorchTensor, tensor)
+
+        raise ValueError(f"Unknown tensor type {type(tensor)}")
+
+    def __init__(self, tensor):
+        self.data = tensor
+
+    def __repr__(self):
+        return self.data.__repr__()
+
+    def __str__(self):
+        return self.data.__str__()
+
+    def __len__(self):
+        return len(self.data)
+
+    def __mul__(self, other):
+        if isinstance(other, UnifiedTensor):
+            other = other.data
+
+        print(self, other)
+        return self.__class__(self.data * other)
+
+    def __rmul__(self, other):
+        if isinstance(other, UnifiedTensor):
+            other = other.data
+
+        return self.__class__(other * self.data)

```

Comment:
I only implemented the special methods currently necessary for now.

We will need to extend this for addition, subtraction, etc.